<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Outlier Detection in Genetic Data • molic</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Outlier Detection in Genetic Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">molic</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/genetic_example.html">Outlier Detection in Genetic Data</a>
    </li>
    <li>
      <a href="../articles/outlier_intro.html">The Outlier Model</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Outlier Detection in Genetic Data</h1>
                        <h4 class="author">Mads Lindskou</h4>
            
      
      
      <div class="hidden name"><code>genetic_example.Rmd</code></div>

    </div>

    
    
<div id="the-1000-genomes-project-data" class="section level1">
<h1 class="hasAnchor">
<a href="#the-1000-genomes-project-data" class="anchor"></a>The 1000 Genomes Project Data</h1>
<p>The outlier detection method <span class="citation">(Lindskou, Svante Eriksen, and Tvedebrink 2019)</span> arose from a problem in the forensic science community where it is of great interest to make statements about the geographical origin of a DNA sample. This is in general a very complicated matter. More over, when DNA markers are in linkage disequilibrium things get even more complicated. The 1000 Genomes Project set out to provide a comprehensive description of common human genetic variation by applying whole-genome sequencing to a diverse set of individuals from multiple populations <span class="citation">(The 1000 Genomes Project Consortium 2015)</span>. In the <strong>molic</strong> package we include the final data from the project which includes <span class="math inline">\(2504\)</span> DNA profiles coming from five different continental regions; Europe (EUR), Africa (AFR), America (AMR), East Asia (EAS), South Asia (SAS). Each DNA profile is the compound of two rows in the data <code>tgp_dat</code>, one for each chromosome copy. This makes the outlier procedure slightly more complicated since we must fit a model to each copy and aggregate the information (but we shall see in a moment that it is not hard to do using the <strong>molic</strong> package).</p>
<p>The data includes <span class="math inline">\(276\)</span> SNP markers grouped in <span class="math inline">\(97\)</span> microhaplotypes; in other words, the SNPs within a microhaplotype are so close that they cannot be assumed to be in linkage disequilibrium and we must take into account their mutal dependencies. We do this with the use of decomposable graphical models. See the <code>outlier_intro</code> for more details on the model. The microhaplotype grouping is provided is the list <code>tgp_haps</code>.</p>
</div>
<div id="analyzing-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-the-data" class="anchor"></a>Analyzing The Data</h1>
<p>We show how to test that an unknown DNA profile <code>z</code> can be assumed to originate from the European database. We shall pick a random DNA profile from the American database and assume we do not know the origin. We first load the necessary libraries</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(molic)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</code></pre></div>
<p>Next, we extract the relevant data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># SNPs grouped into (micro) haplotypes</span>
haps &lt;-<span class="st"> </span>tgp_haps

<span class="co"># All the Europeans</span>
eur  &lt;-<span class="st"> </span>tgp_dat <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(pop_meta <span class="op">==</span><span class="st"> "EUR"</span>)

<span class="co"># Extracting the two databases for each copy of the chromosomes</span>
eur_a &lt;-<span class="st"> </span>eur <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"a$"</span>, sample_name))

eur_b &lt;-<span class="st"> </span>eur <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"b$"</span>, sample_name))</code></pre></div>
<p>Then we need to model the relationship between SNPs within all chromosomes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fitting graphs for the two chromosome copies</span>
Ga &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(haps, <span class="cf">function</span>(h) {
  <span class="kw"><a href="../reference/fit_graph.html">fit_graph</a></span>(eur_a <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(h), <span class="dt">trace =</span> <span class="ot">FALSE</span>)
})

Gb &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(haps, <span class="cf">function</span>(h) {
  <span class="kw"><a href="../reference/fit_graph.html">fit_graph</a></span>(eur_b <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(h), <span class="dt">trace =</span> <span class="ot">FALSE</span>)
})</code></pre></div>
<p>The variables <code>Ga</code> and <code>Gb</code> are now <code>list</code> objects where each object is a <code><a href="https://www.rdocumentation.org/packages/base/topics/c">c("fwd", "gengraph", "list")</a></code> object. Lets print and plot one of them,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(Ga[[<span class="dv">2</span>]])</code></pre></div>
<pre><code>##  A Decomposable Graph With 
##  ------------------------- 
##   Nodes: 3 
##   Edges: 1 / 3 
##   Cliques: 2 
##   &lt;fwd, gengraph, list&gt; 
##  -------------------------</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(Ga[[<span class="dv">1</span>]])</code></pre></div>
<p><img src="genetic_example_files/figure-html/b-1.png" width="700" style="display: block; margin: auto;"></p>
<p>But all of these individual graphs should be represented as one big graph</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Merge the fitted graphs into a single graph</span>
Ga_union &lt;-<span class="st"> </span>Ga <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="cf">function</span>(x) <span class="kw"><a href="../reference/adj_lst.html">adj_lst</a></span>(x)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(<span class="dt">recursive =</span> <span class="ot">FALSE</span>)

Gb_union &lt;-<span class="st"> </span>Gb <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="cf">function</span>(x) <span class="kw"><a href="../reference/adj_lst.html">adj_lst</a></span>(x)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(<span class="dt">recursive =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>These objects are now plain <code>list</code> objects. But we can convert them to a <code>gengraph</code> in order to plot and print</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We can convert the adjacency lists to a gengraph object and plot/print the graph</span>
g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gengraph.html">gengraph</a></span>(eur_a <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(haps))), <span class="st">"gen"</span>, Ga_union)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(g)</code></pre></div>
<pre><code>##  A Decomposable Graph With 
##  ------------------------- 
##   Nodes: 302 
##   Edges: 234 / 45451 
##   Cliques: 161 
##   &lt;gengraph, list&gt; 
##  -------------------------</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(g, <span class="dt">vertex.size =</span> <span class="dv">1</span>, <span class="dt">vertex.label.cex =</span> <span class="fl">0.1</span>)</code></pre></div>
<p><img src="genetic_example_files/figure-html/d-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Next, we extract a DNA profile from the American database</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">amr &lt;-<span class="st"> </span>tgp_dat <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(pop_meta <span class="op">==</span><span class="st"> "AMR"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">20183007</span>)
z1  &lt;-<span class="st"> </span>amr <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"a$"</span>, sample_name)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(haps))) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/sample.html">sample_n</a></span>(<span class="dt">size =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>()

z2  &lt;-<span class="st"> </span>amr <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"b$"</span>, sample_name)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(haps))) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/sample.html">sample_n</a></span>(<span class="dt">size =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>()</code></pre></div>
<p>and since the hypothesis is that <code>z</code> belongs to <code>eur</code>, we append <code>z</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dz1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(eur_a, z1)
dz2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(eur_b, z2)</code></pre></div>
<p>We are now ready to fit the outlier model</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ma &lt;-<span class="st"> </span><span class="kw"><a href="../reference/outlier_model.html">outlier_model</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(dz1), Ga_union, <span class="dt">nsim =</span> <span class="dv">5000</span>, <span class="dt">ncores =</span> <span class="dv">1</span>, <span class="dt">validate =</span> <span class="ot">FALSE</span>)
mb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/outlier_model.html">outlier_model</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(dz2), Gb_union, <span class="dt">nsim =</span> <span class="dv">5000</span>, <span class="dt">ncores =</span> <span class="dv">1</span>, <span class="dt">validate =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>And finally, we combine the information into a single mixed outlier test</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fit_mixed_outlier.html">fit_mixed_outlier</a></span>(ma, mb, z1, z2)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(m)</code></pre></div>
<pre><code>## 
##  -------------------------------- 
##   Simulations: 5000 
##   Variables: 304 
##   Observations: 1008 
##   Estimated mean: -1459.47 
##   Estimated variance: 96.92 
##     ---------------------------   
##   Critical value: -1442.477 
##   Deviance: -1400.392 
##   P-value: 0 
##   Alpha: 0.05 
##   &lt;mixed_outlier, outlier, outlier_model, list&gt; 
##  --------------------------------</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualize the density of the deviance statistic</span>
<span class="kw"><a href="../reference/pmf.html">pmf</a></span>(m)</code></pre></div>
<p><img src="genetic_example_files/figure-html/h-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can conclude, from the print and pmf, that <code>z</code> does not originate from Europe with a high degree of certainty.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-lindskououtlier">
<p>Lindskou, Mads, Poul Svante Eriksen, and Torben Tvedebrink. 2019. “Outlier Detection in Contingency Tables Using Decomposable Graphical Models.” <em>Scandinavian Journal of Statistics</em>. Wiley Online Library. doi:<a href="https://doi.org/10.1111/sjos.12407">10.1111/sjos.12407</a>.</p>
</div>
<div id="ref-10002015global">
<p>The 1000 Genomes Project Consortium. 2015. “A Global Reference for Human Genetic Variation.” <em>Nature</em> 526 (7571). Nature Publishing Group: 68. doi:<a href="https://doi.org/10.1038/nature15393">10.1038/nature15393</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#the-1000-genomes-project-data">The 1000 Genomes Project Data</a></li>
      <li><a href="#analyzing-the-data">Analyzing The Data</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Mads Lindskou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
